src/transformStateWithClones.js

```
@@ -6,8 +6,36 @@
  *
  * @return {Object[]}
  */
-function transformStateWithClones(state, actions) {
-  // write code here
+function transformStateWithClones(initialState, actions) {
+  let currentState = { ...initialState };
+  const history = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        currentState = Object.fromEntries(
+          Object.entries(currentState).filter(
+            ([key]) => !action.keysToRemove.includes(key),
+          ),
+        );
+        break;
+
+      default:
+        break;
```

This violates checklist item #3: 'switch/case should always have default case for error handling'. The current `default` case is empty, which means any unknown action types will be silently ignored. To make the function more robust, consider throwing an error here to indicate that an invalid action was provided.